File: lib-segment/src/main/java/websiteschema/mpsegment/graph/WordBigram.java
package websiteschema.mpsegment.graph;

import websiteschema.mpsegment.hmm.Node;
import websiteschema.mpsegment.hmm.NodeRepository;
import websiteschema.mpsegment.hmm.Trie;
import websiteschema.mpsegment.util.SerializeHandler;

import java.io.*;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class WordBigram {

    private static final Map<String,WordBigram> mapWordBigram = new ConcurrentHashMap<String, WordBigram>();
    private Trie trie = new Trie();
    private NodeRepository nodeRepository = new NodeRepository();

    public WordBigram() {
    }

    public WordBigram(Trie trie, NodeRepository nodeRepository) {
        this.trie = trie;
        this.nodeRepository = nodeRepository;
    }

    public WordBigram(String resource) {
        try {
            load(resource);
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    public static WordBigram getInstance(String corpusName) {
        return mapWordBigram.containsKey(corpusName) ? mapWordBigram.get(corpusName) : createWordBigram(corpusName);
    }

    private static WordBigram createWordBigram(String corpusName) {
        WordBigram instance;
        synchronized (mapWordBigram) {
            if(!mapWordBigram.containsKey(corpusName)) {
                instance = new WordBigram(corpusName);
                mapWordBigram.put(corpusName, instance);
            } else {
                instance = mapWordBigram.get(corpusName);
            }
        }
        return instance;
    }

    public void load(String resource) throws IOException {
        InputStream inputStream = getInputStream(resource);
        SerializeHandler reader = new SerializeHandler(new DataInputStream(inputStream));
        nodeRepository.load(reader);
        trie.load(reader);
        trie.buildIndex(1);
    }

    public double getProbability(String word1, String word2) {
        Node firstWord = nodeRepository.get(getWordFirstTwoChars(word1));
        Node secondWord = nodeRepository.get(getWordFirstTwoChars(word2));
        if(firstWord == null || secondWord == null) {
            return -1;
        }
        int[] bigram = new int[]{
                firstWord.getIndex(),
                secondWord.getIndex()};
        Trie node = trie.searchNode(bigram);
        return null != node ? node.getProb() : -1;
    }

    private InputStream getInputStream(String resource) throws FileNotFoundException {
        InputStream inputStream = null;
        File file = new File(resource);
        if (file.exists()) {
            inputStream = new FileInputStream(file);
        } else {
            inputStream = getClass().getClassLoader().getResourceAsStream(resource);
        }
        return inputStream;
    }

    private String getWordFirstTwoChars(String word) {
        return word.length() > 2 ? word.substring(word.length() - 2) : word;
    }
}

File: lib-segment/src/main/scala/websiteschema/mpsegment/graph/WordBigram.scala
package websiteschema.mpsegment.graph

import websiteschema.mpsegment.hmm.NodeRepository
import websiteschema.mpsegment.hmm.Trie
import websiteschema.mpsegment.util.SerializeHandler
import collection.immutable.HashMap
import java.io._

class WordBigram {

  private var trie: Trie = new Trie()
  private var nodeRepository: NodeRepository = new NodeRepository()

  private def load(resource: String) {
    val inputStream = getInputStream(resource)
    val reader = SerializeHandler(new DataInputStream(inputStream))
    nodeRepository.load(reader)
    trie.load(reader)
    trie.buildIndex(1)
  }

  def getProbability(word1: String, word2: String): Double = {
    val firstWord = nodeRepository.get(getWordFirstTwoChars(word1))
    val secondWord = nodeRepository.get(getWordFirstTwoChars(word2))
    if (firstWord == null || secondWord == null) {
      return -1
    }
    val bigram = Array[Int](
      firstWord.getIndex(),
      secondWord.getIndex()
    )
    val node = trie.searchNode(bigram)
    if(null != node) node.getProb() else -1.0D
  }

  private def getInputStream(resource: String): InputStream = {
    var inputStream: InputStream = null
    val file = new File(resource)
    if (file.exists()) {
      inputStream = new FileInputStream(file)
    } else {
      inputStream = getClass().getClassLoader().getResourceAsStream(resource)
    }
    return inputStream
  }

  private def getWordFirstTwoChars(word: String): String = {
    return if(word.length() > 2) word.substring(word.length() - 2) else word
  }
}

object WordBigram {
  private var mapWordBigram = HashMap[String, WordBigram]()

  def apply(corpusName: String) = {
    createWordBigram(corpusName)
  }

  def apply(trie: Trie, nodeRepository: NodeRepository) = {
    val bigram: WordBigram = new WordBigram()
    bigram.trie = trie
    bigram.nodeRepository = nodeRepository
    bigram
  }

  private def createWordBigram(corpusName: String): WordBigram = {
    if (!mapWordBigram.contains(corpusName)) {
      val instance = new WordBigram()
      instance.load(corpusName)
      mapWordBigram += (corpusName -> instance)
      instance
    } else {
      mapWordBigram(corpusName)
    }
  }
}
